### PR TITLE
GH-740 & GH-741 Table Formatting

### PR DESCRIPTION
**Pull request by dtsnell4**
Adds formatting for numbers, strings and dates, including publish parameters for right/left/center justification for each type.

Fixes the extra row when there is no "total"

Fixes GH-740
Fixes GH-741
Fixes GH-1143
Fixes GH-1162
Fixes GH-1165

Signed-off-by: Dan Snell <Dan.Snell@lexisnexis.com>

### CODE REVIEW

**Comment by dtsnell4** [comment]
@GordonSmith @mlzummo @jbrundage Please review

http://rawgit.com/dtsnell4/Visualization/GH-740_Table_Formatting/demos/dermatology.html?src/other/Table.formatted

**Comment by mlzummo** [comment]
you need to fetch before doing a rebease ... which is why your geting lint errors

**Comment by mlzummo** [comment]
looks good

**Comment by dtsnell4** [comment]
@GordonSmith This also fixes bower.json which was pointing to the wrong version of amcharts

File: bower.json [hunk 1]
```diff
@@ -42,7 +42,7 @@
     "topojson": "~1.6.19",
     "dagre": "https://raw.githubusercontent.com/cpettitt/dagre/v0.7.1/dist/dagre.js",
     "d3-cloud": "~1.2.0",
-    "amcharts3": "~3.17.0",
+    "amcharts3": "~3.17.3",
```
Comment (GordonSmith) [inline bower.json]: This change should not be in this PR

File: src/other/Table.js [hunk 2]
```diff
@@ -669,48 +695,52 @@
...
-    Table.prototype.applyFirstColRowStyles = function(rows){
-        this.applyStyleToRows(rows,true);
+        if(typeof(cellData) === "string"){
+            var isDate = Date.parse(cellData);
+            if (isNaN(isDate)) {
+                return cellData.trim();
+            } else {
+                var timeFormat = d3.time.format(context.columnPatterns()[colIdx]);
+                return timeFormat(new Date(cellData)) || cellData;
+            }
```
Comment (GordonSmith) [inline src/other/Table.js]: You are mixing up data parsing and presentation formatting here?

File: test/otherFactory.js [hunk 3]
```diff
@@ -63,6 +63,16 @@
                     );
                 });
             },
+            formatted: function (callback) {
+                require(["test/DataFactory", "src/other/Table"], function (DataFactory, Table) {
+                    callback(new Table()
+                        .columns(DataFactory.Table.formatted.columns)
+                        .data(DataFactory.Table.formatted.data)
+                        .columnPatterns([".4%", ".6r", "%d-%m-%y"])
```
Comment (GordonSmith) [inline test/otherFactory.js]: Can you add a test which has empty elements - say I only want to format the 3rd column.

**Comment by GordonSmith** [comment]
@dtsnell4 You should not be doing any data parsing (specifically the conversion of a String->Date) that should be done as a separate exercise (and is something which the new DB engine will encapsulate well).  It also ties in with the date patterns we use in the charts.

IOW you should simply be just formatting the data as is with the specified format string, if it fails then the user is applying an invalid format string to the given column.

File: src/other/Table.js (hunk 2 above)
Comment (dtsnell4) [inline src/other/Table.js]: I do not believe so, can you elaborate?

File: test/otherFactory.js (hunk 3 above)
Comment (dtsnell4) [inline test/otherFactory.js]: Yes, I need a better sample data set, but if you want to test quickly you can just edit the array in the prop editor

**Comment by dtsnell4** [comment]
@GordonSmith updated per comments

File: src/other/Table.js [hunk 4]
```diff
@@ -669,48 +695,47 @@
...
-                if (tr.classed("hover")) {
-                    context.applyHoverRowStyles(tr);
-                } else if (tr.classed("selected")) {
-                    context.applySelectedRowStyles(tr);
-                } else {
-                    context.applyRowStyles(tr,isFirstCol);
-                }
-            })
-        ;
+        switch(typeof(cellData)){
```
Comment (mlzummo) [inline src/other/Table.js]: more than likely this block right here will be replaced by calls to the new DB object

File: src/other/Table.js [hunk 5]
```diff
@@ -669,48 +695,47 @@
...
-    };
-    Table.prototype.applyRowStyles = function(row, isFirstCol){
-        row
-            .style("color", isFirstCol ? this.tbodyFirstColFontColor() : this.tbodyFontColor())
-            .style("background-color", isFirstCol ? this.tbodyFirstColBackgroundColor() : this.tableZebraColor_exists() && this.data().indexOf(row.datum()) % 2 ? this.tbodyRowBackgroundColor() : this.tableZebraColor())
-        ;
-    };
-    Table.prototype.applyFirstColRowStyles = function(rows){
-        this.applyStyleToRows(rows,true);
+        if(typeof(cellData) === "string"){
```
Comment (mlzummo) [inline src/other/Table.js]: more than likely this block right here will be replaced by calls to the new DB object

**Comment by GordonSmith** [comment]
@dtsnell4 Needs rebase...

**Comment by dtsnell4** [comment]
@GordonSmith Finally finished rebase

**Comment by dtsnell4** [comment]
For what its worth, I've had a bit of an epiphany on this Table widget, and I believe I now know how to refactor it to get rid of the "copying" code thats making it resource intensive and make it more "d3-like", as well as cut out a lot of the other dom manipulation code.  I'm not sure how far Gordon got with this, but I will look into as time permits.

**Comment by GordonSmith** [comment]
@dtsnell4 Can you update the fixes references in the commit message to include all the issues it is now fixing?

Also I think this needs a rebase to get rid of the travis errors.

**Comment by dtsnell4** [comment]
I rebased it just yesterday, but I'll try again

**Comment by dtsnell4** [comment]
@GordonSmith Rebased

**Comment by dtsnell4** [comment]
@GordonSmith Matt helped me figure out what was wrong with  Travis errors.
